patches/tModLoader/Terraria/ModLoader/BiomeLoader.cs

```diff
@@ -0,0 +1,142 @@
+
+		private static HookList AddHook<F>(Expression<Func<ModBiome, F>> func) {
+			var hook = new HookList(ModLoader.Method(func));
+			hooks.Add(hook);
+			return hook;
+		}
+
+		internal static void Add(ModBiome biome) {
+			biome.index = biomes.Count;
+			biomes.Add(biome);
+		}
+
+		internal static void RebuildHooks() {
+			foreach (var hook in hooks) {
+				hook.arr = ModLoader.BuildGlobalHook(biomes, hook.method).Select(p => p.index).ToArray();
+			}
+		}
+
+		internal static void CopyBiomesTo(Player newPlayer) {
+			newPlayer.modBiomes = new ModBiome[biomes.Count];
+			biomes.CopyTo(newPlayer.modBiomes, 0);
+		}
+
+		public static void UpdateBiomes(Player player) {
+			foreach (var biome in player.modBiomes) {
+				biome.Update();
+			}
+		}
+
+		public static bool CustomBiomesMatch(Player player, Player other) {
+			for (int i = 0; i < player.modBiomes.Length; i++) {
+				if (player.modBiomes[i].Active ^ other.modBiomes[i].Active)
+					return false;
+			}
+			return true;
+		}
+
+		public static void CopyCustomBiomesTo(Player player, Player other) {
+			for (int i = 0; i < player.modBiomes.Length; i++) {
+				other.modBiomes[i].Active = player.modBiomes[i].Active;
+			}
+		}
+
+		public static void SendCustomBiomes(Player player, BinaryWriter writer) {
+			ushort count = 0;
+			byte[] data;
+			using (MemoryStream stream = new MemoryStream()) {
+				using BinaryWriter customWriter = new BinaryWriter(stream);
+				for (int i = 0; i < player.modBiomes.Length; i++) {
+					if (SendCustomBiomes(player.modBiomes[i], customWriter)) {
+						count++;
+					}
+				}
+				customWriter.Flush();
+				data = stream.ToArray();
+			}
+			writer.Write(count);
+			writer.Write(data);
+		}
+
+		private static bool SendCustomBiomes(ModBiome modBiome, BinaryWriter writer) {
+			byte[] data;
+			using (MemoryStream stream = new MemoryStream()) {
+				using BinaryWriter customWriter = new BinaryWriter(stream);
+				//modPlayer.SendCustomBiomes(writer);
+				customWriter.Write(modBiome.Active);
+				customWriter.Flush();
+				data = stream.ToArray();
+			}
+			if (data.Length > 0) {
+				writer.Write(modBiome.Mod.Name);
+				writer.Write(modBiome.Name);
+				writer.Write((byte)data.Length);
+				writer.Write(data);
+				return true;
+			}
+			return false;
+		}
+
+		public static void ReceiveCustomBiomes(Player player, BinaryReader reader) {
+			int count = reader.ReadUInt16();
+
+			for (int k = 0; k < count; k++) {
+				string modName = reader.ReadString();
+				string name = reader.ReadString();
+				byte[] data = reader.ReadBytes(reader.ReadByte());
+
+				if (ModContent.TryFind<ModBiome>(modName, name, out var modBiomeBase)) {
+					var modBiome = player.GetModBiome(modBiomeBase);
+
+					using MemoryStream stream = new MemoryStream(data);
+					using BinaryReader customReader = new BinaryReader(stream);
+
+					modBiome.Active = customReader.ReadBoolean();
+					//try { modPlayer.ReceiveCustomBiomes(reader) }
+				}
```

Comment:
Again, this stuff with `BinaryWriter(MemoryStream)` is now going to cause overhead. I'll fix that in a bit.